forte/data/extractor/attribute_extractor.py

```diff
@@ -0,0 +1,189 @@
+# Copyright 2020 The Forte Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import logging
+from collections import Hashable, abc
+from typing import Dict, Any, Union, Iterable
+from ft.onto.base_ontology import Entry, Annotation
+from forte.common.configuration import Config
+from forte.data.data_pack import DataPack
+from forte.data.converter.feature import Feature
+from forte.data.extractor.base_extractor import BaseExtractor
+
+logger = logging.getLogger(__name__)
+
+__all__ = [
+    "AttributeExtractor"
+]
+
+
+class AttributeExtractor(BaseExtractor):
```

Comment:
One thing I forgot to mention.

For future parallel programming purposes. We need to reduce the amount of internal states. One thing you notice that is shared across the whole pipeline is `Resources` in the processor initialize. If we put all shared states to `Resources` that would be much easier for us to parallel them in the future.

Could you pass in the `Resources` class here? And put the `self.vocab` from Extractor into `Resources`? However, you need to use a specific key when you are adding to the resource. It cannot be the class name because there may be multiple `attribute_extractor`.

Another option is to make the object `Vocabulary` parallel in the future.

Reply:
Extractor currently is not a Processor, so there is no Resource and there is no initialize method. And I am not sure whether it is compatible to processor right now.

I take a look at Resource class. If we put self.vocab into resources directly, it will be saving the class object directly, which might conflict with our current design, turning class object into state and save the state. 

Can you be a little bit more specific how we are going to use this Resource class? For example,
1. create a resource class in __init__ or receive a resource class from outside
2. put self.vocab into resource or self.vocab.state into resource
3. what is the relation between this resource class and the state, from_state function?

Reply:
1. You don't have to make these Processors, you can simply pass a `Resources` from a processor.
2. We shouldn't create our own name `from_state`, `state`. Instead, if we implement `__getstate__`, `__setstate__` correctly, the pickle function will work automatically. [get_state](https://docs.python.org/3/library/pickle.html#object.__getstate__
). So you can put `self.vocab` to resources, and it will be load back automatically. Note that these are not classmethod, but instance method.

Reply:
Take a look at DataPack serialisation as example:

https://github.com/asyml/forte/blob/master/forte/data/data_pack.py#L86

Reply:
I see. This seems to be a better solution. I will look into the __setstate__ and __getstate__ and update my implementation.